File: src/main/java/com/github/morningzeng/toolset/ui/coding/AbstractCodingComponent.java
package com.github.morningzeng.toolset.ui.coding;

import com.github.morningzeng.toolset.Constants.IconC;
import com.github.morningzeng.toolset.component.LanguageTextArea;
import com.github.morningzeng.toolset.utils.GridBagUtils;
import com.github.morningzeng.toolset.utils.GridBagUtils.GridBagFill;
import com.intellij.openapi.fileTypes.PlainTextLanguage;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import com.intellij.ui.components.JBPanel;
import com.intellij.ui.components.JBPanelWithEmptyText;

import javax.swing.JButton;
import java.util.function.Supplier;

/**
 * @author Morning Zeng
 * @since 2024-11-12
 */
sealed abstract class AbstractCodingComponent extends JBPanel<JBPanelWithEmptyText> permits Base64Component, URLComponent {

    final LanguageTextArea encodeArea;
    final LanguageTextArea decodeArea;
    private final JButton encodeBtn = new JButton("Encode", IconC.DOUBLE_ANGLES_DOWN);
    private final JButton decodeBtn = new JButton("Decode", IconC.DOUBLE_ANGLES_UP);

    AbstractCodingComponent(final Project project) {
        this.encodeArea = new LanguageTextArea(PlainTextLanguage.INSTANCE, project, "");
        this.decodeArea = new LanguageTextArea(PlainTextLanguage.INSTANCE, project, "");

        this.initLayout();
        this.initEvent();
    }

    void initLayout() {
        GridBagUtils.builder(this)
                .fill(GridBagFill.BOTH)
                .newRow(row -> row.newCell().weightX(1).weightY(1).add(this.decodeArea.withRightBar()))
                .newRow(row -> {
                    final JBPanel<JBPanelWithEmptyText> btnPanel = GridBagUtils.builder()
                            .fill(GridBagFill.HORIZONTAL)
                            .newRow(_row -> _row.newCell().add(this.encodeBtn)
                                    .newCell().add(this.decodeBtn))
                            .build();
                    row.fill(GridBagFill.HORIZONTAL).newCell().weightY(0).add(btnPanel);
                })
                .fill(GridBagFill.BOTH)
                .newRow(row -> row.newCell().weightY(1).add(this.encodeArea.withRightBar()));
    }

    void initEvent() {
        this.encodeBtn.addActionListener(e -> {
            try {
                final String enc = this.encode().get();
                this.encodeArea.setText(enc);
            } catch (Exception ex) {
                Messages.showMessageDialog(this, ex.getMessage(), "Encoding Error", Messages.getErrorIcon());
            }
        });
        this.decodeBtn.addActionListener(e -> {
            try {
                final String dec = this.decode().get();
                this.decodeArea.setText(dec);
            } catch (Exception ex) {
                Messages.showMessageDialog(this, ex.getMessage(), "Decoding Error", Messages.getErrorIcon());
            }
        });
    }

    abstract Supplier<String> encode();

    abstract Supplier<String> decode();

}

File: src/main/kotlin/com/github/morningzeng/toolbox/ui/view/coding/AbstractCodingComponent.kt
package com.github.morningzeng.toolbox.ui.view.coding

import com.github.morningzeng.toolbox.Constants
import com.github.morningzeng.toolbox.ui.action.HistoryAction
import com.github.morningzeng.toolbox.ui.component.LanguageTextArea
import com.github.morningzeng.toolbox.utils.GridBagUtils
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.ui.components.JBPanel
import com.intellij.ui.components.JBPanelWithEmptyText
import javax.swing.JButton

/**
 * @author Morning Zeng
 * @since 2025-07-22
 */
abstract class AbstractCodingComponent(
    protected val project: Project
) : JBPanel<JBPanelWithEmptyText>() {

    private val encHistoryAction: HistoryAction<String> = HistoryAction()
    private val decHistoryAction: HistoryAction<String> = HistoryAction()
    val encodeArea: LanguageTextArea = textArea(encHistoryAction)
    val decodeArea: LanguageTextArea = textArea(decHistoryAction)
    private val encodeBtn: JButton = JButton("Encode", Constants.IconC.DOUBLE_ANGLES_DOWN)
    private val decodeBtn: JButton = JButton("Decode", Constants.IconC.DOUBLE_ANGLES_UP)

    init {
        initLayout()
        initEvent()
    }

    fun initLayout() {
        GridBagUtils.builder(this)
            .fill(GridBagUtils.GridBagFill.BOTH)
            .row { it.cell().weightX(1.0).weightY(1.0).add(decodeArea.withRightBar()) }
            .row {
                val panel = GridBagUtils.builder().fill(GridBagUtils.GridBagFill.HORIZONTAL)
                    .row { cit ->
                        cit.cell().add(encodeBtn)
                            .cell().add(decodeBtn)
                    }
                    .build()
                it.fill(GridBagUtils.GridBagFill.HORIZONTAL).cell().weightY(0.0).add(panel)
            }
            .fill(GridBagUtils.GridBagFill.BOTH)
            .row { it.cell().weightY(1.0).add(encodeArea.withRightBar()) }
    }

    fun initEvent() {
        encodeBtn.addActionListener {
            try {
                encodeArea.text = encode()
                encHistoryAction.addItem(textConvert()(""))
                decHistoryAction.addItem(textConvert()(""))
            } catch (e: Exception) {
                Messages.showErrorDialog(project, e.message, "Encode Error")
            }
        }
        decodeBtn.addActionListener {
            try {
                decodeArea.text = decode()
                encHistoryAction.addItem(textConvert()(""))
                decHistoryAction.addItem(textConvert()(""))
            } catch (e: Exception) {
                Messages.showErrorDialog(project, e.message, "Decode Error")
            }
        }
    }

    fun textConvert(): (String) -> String {
        return { "${decodeArea.text} - ${encodeArea.text}" }
    }

    fun tConvert(): (String) -> String? {
        return {
            it.split(" - ").apply {
                decodeArea.text = this[0]
                encodeArea.text = this[1]
            }
            null
        }
    }

    abstract fun encode(): String

    abstract fun decode(): String

    private fun textArea(action: HistoryAction<String>): LanguageTextArea {
        return object : LanguageTextArea(project, "") {
            override fun defaultRightBarActions(): Array<AnAction> {
                return arrayOf(
                    softWrapAction(),
                    scrollToEndAction(),
                    action.also {
                        it.setTextArea(this)
                        it.textAreaEvent(textConvert(), tConvert())
                    },
                    clearAllAction()
                )
            }
        }
    }

}